sql/7_0_3-to-7_0_4_upgrade.sql

```diff
@@ -120,3 +120,19 @@ ALTER TABLE `onetime_auth` ADD `profile` tinytext COMMENT 'profile of scope for
 #IfMissingColumn onetime_auth onetime_actions
 ALTER TABLE `onetime_auth` ADD `onetime_actions` text COMMENT 'JSON array of actions that can be performed with this token';
 #EndIf
+
+#ifNotTable track_events
+CREATE TABLE `track_events`
+(
+    `id`           BIGINT(12) UNSIGNED NOT NULL AUTO_INCREMENT,
+    `event_type`   VARCHAR(50)  NOT NULL,
+    `event_label`  VARCHAR(255) NOT NULL,
+    `event_url`    VARCHAR(255) DEFAULT NULL,
+    `event_target` VARCHAR(50)  DEFAULT NULL,
+    `first_event`   DATETIME   NOT NULL,
```

Comment:
Are first_event,last_event and count a measure to save space?

From an architecture view will we want to track larger amounts of data that won't fit within a 255 varchar?

I'm thinking if we want to expand events at some point in the futureto include instance installation details such as (language, region, timezone, portal_enabled) etc?

Reply:
For click events I only care that it was clicked. Didn't want to recorded new row for every click with turns into duplicated event records. The eventType would give use event context like portal or module etc.
The AWS DynamoDB stores the events array as json object so structure is free to change.


> I'm thinking if we want to expand events at some point in the futureto include instance installation details such as (language, region, timezone, portal_enabled) etc?

These are normalized and indexed. We can add new columns dynamically as NoSQL db.

Reply:
Since the json structure may change in the future, does it makes sense to version it (ie. dynamodb column with json_version (or some other name) that stores the version of the json (so can then make changes and be backward compatible to still do analytics on prior version json structures)?